### ISSUE TITLE
Is Pester thread safe? Or bug in 4.5.0-beta1?

### ISSUE THREAD

**Issue by bergmeister** [pester/Pester#1181]
I was running a long `Pester` run in the background and decided to run another suite (in another PowerShell session) a few times (to see if/why one test suite sporadically fails). I sometimes got this error and was wondering if Pester is supposed to be thread safe? My test suite was not using anything like BeforeEach, BeforeAll, AfterEach, AfterAll,mocking, etc.
I am using PowerShell 5.1 and Pester 4.5.0-beta1:

```
Get-PSDrive : Cannot find drive. A drive with the name 'TestRegistry' does not exist.
At C:\Users\cberg\Documents\WindowsPowerShell\Modules\Pester\4.5.0\Functions\TestRegistry.ps1:61 char:47
+ ... egistry::" + (& $SafeCommands['Get-PSDrive'] -Name TestRegistry).Root
+                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (TestRegistry:String) [Get-PSDrive], DriveNotFoundException
    + FullyQualifiedErrorId : GetLocationNoMatchingDrive,Microsoft.PowerShell.Commands.GetPSDriveCommand

Remove-Item : The registry key at the specified path does not exist.
At C:\Users\cberg\Documents\WindowsPowerShell\Modules\Pester\4.5.0\Functions\TestRegistry.ps1:79 char:13
+             & $SafeCommands['Remove-Item'] -Force -Recurse
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OpenError: (:String) [Remove-Item], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.RemoveItemCommand
```

**Comment by nohwnd** [comment]
The drive is set to Global and always was, so it should be process isolated but not thread safe. This is a new feature so I might make some error when I fixed the implementation in the pull request.

**Comment by bergmeister** [comment]
Ok, thanks for the info

**Comment by nohwnd** [comment]
Okay think I found it. First a simple repro: 

```powershell
Get-Module Pester | Remove-Module
Import-Module C:\Projects\pester_main\Pester.psd1

while ($true) {
    try {
    Describe "A" { It "b" {}}
    } catch { 
    $global:err = $_ 
    throw $_
    }
}
```
Run this from two powershell sessions at the same time and wait. If you get no errors then try stopping and starting one of them, to trigger this issue the timing must be offset by some non-determined margin. Usually it does not take long to get failures.

The errors that you reported are a by-product of the logic, which imho all boils down to a race condition in `New-Item -Force`, as this rather cryptic error suggests, one of the instances is trying to delete the Pester registry folder while the other one just deleted it, as a result we get this error: 

```
New-Item : Cannot delete a subkey tree because the subkey does not exist.
At C:\Projects\pester_main\Functions\Environment.ps1:48 char:5
+     New-Item -Path 'HKCU:\Software\Pester' -Force
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : WriteError: (HKEY_CURRENT_USER\Software\Pester:String) [New-Item], ArgumentException
+ FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.NewItemCommand
```

The Pester folder should be created when first needed and then stay in place, because we put the Guid folders in it so recreating it for every test does not make sense anyway, so to solve this simply removing the `-Force` should be enough. I tested it and it seems to work just fine afterwards.

If you wonder why this does not happen with TestDrive, it is because there we are using the already existing Temp folder, so no contention exists there.
